coderd/tailnet.go

```diff
@@ -537,13 +539,20 @@ func NewMultiAgentController(ctx context.Context, logger slog.Logger, tracer tra
 // InmemTailnetDialer is a tailnet.ControlProtocolDialer that connects to a Coordinator and DERPMap
 // service running in the same memory space.
 type InmemTailnetDialer struct {
-	CoordPtr *atomic.Pointer[tailnet.Coordinator]
-	DERPFn   func() *tailcfg.DERPMap
-	Logger   slog.Logger
-	ClientID uuid.UUID
+	CoordPtr              *atomic.Pointer[tailnet.Coordinator]
+	DERPFn                func() *tailcfg.DERPMap
+	Logger                slog.Logger
+	ClientID              uuid.UUID
+	DatabaseHealthcheckFn func(ctx context.Context) error
 }
 
-func (a *InmemTailnetDialer) Dial(_ context.Context, _ tailnet.ResumeTokenController) (tailnet.ControlProtocolClients, error) {
+func (a *InmemTailnetDialer) Dial(ctx context.Context, _ tailnet.ResumeTokenController) (tailnet.ControlProtocolClients, error) {
+	if a.DatabaseHealthcheckFn != nil {
+		if err := a.DatabaseHealthcheckFn(ctx); err != nil {
+			return tailnet.ControlProtocolClients{}, xerrors.Errorf("%s: %w", codersdk.DatabaseNotReachable, err)
+		}
+	}
```

Comment:
I think we need to debounce this. I'm not sure how often we're going to be dialing but my gut tells me that if we call the healthcheck on each ping _synchronously_, we're going to be

1) adding latency to the dialer,
2) adding unwanted load to the database